### ISSUE TITLE
exp(p3): COMA — counterfactual multi-agent credit assignment

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#284]
## Thesis context

Per the thesis, the per-step advantage estimator in our game is biased toward defection because individual contributions to team rewards are diluted across coordinators. **COMA (Foerster et al. 2018)** addresses this directly: it replaces the standard advantage with a counterfactual `A_i = Q(s, a) − Σ_{a_i'} π(a_i'|s) Q(s, (a_-i, a_i'))`, isolating agent i's marginal contribution.

This is conspicuously absent from our 5 failed interventions. We tried MAPPO (centralized critic shared across actors; PR #232 / issue #208/#231); we did NOT try COMA (counterfactual baseline that isolates per-agent credit).

## Protocol

1. Implement COMA in `bucket_brigade/training/` — adapted from public reference implementations (PyMARL, EPyMARL)
2. 50-iter × 3 seeds smoke on `minimal_specialization` and `default`
3. Compare gap_closed to MAPPO baseline (PR #232: tier 3)

## Verdict logic

| Outcome | Interpretation |
|---|---|
| COMA gap_closed ≥ 0.5 | Counterfactual credit assignment solves the credit-dilution part of the misalignment. |
| Similar to MAPPO (tier 3) | Credit assignment isn't the bottleneck; the misalignment is elsewhere. |
| Worse than IPPO | COMA's counterfactual baseline is high-variance; may need variance reduction. |

## Effort

~3 days implementation. Significant — Q-network architecture changes. Worth the investment given thesis centrality.

## Sequencing

Run AFTER #270. If basin trap, lower priority (BC-init solves it). If anti-attractor, COMA is one of the most-on-thesis algorithmic candidates.

## References

- Foerster et al. (2018) "Counterfactual Multi-Agent Policy Gradients" (AAAI)
- `research_notebook/2026-05-17_thesis_misaligned_gradients.md`
- `research_notebook/2026-05-17_lit_review.md`
- PR #232 (MAPPO verdict — shared critic didn't help; counterfactual baseline might)
- Reference implementations: PyMARL (`oxwhirl/pymarl`), EPyMARL (`uoe-agents/epymarl`)

---

### Priority note (per author comment on this issue)

After #271 falsified the active-anti-attractor reading, the user has explicitly **de-prioritised COMA**: the new thesis frame is "gradient locality, not direction," which reduces COMA's expected leverage. Per the comment thread, COMA is now **Phase-3 contingent** — only schedule if the cheaper basin-escape experiments (PBT #288, macro-actions #286) don't help. Reference: `research_notebook/2026-05-17_issue271_best_of_n_verdict.md`. Builder should hold off claiming until human re-promotes; this curation is documentation, not a clear-to-proceed.

### COMA algorithm — math and design choices

COMA = centralized **Q-critic** + per-agent **decentralized actors** (CTDE), with a per-agent counterfactual advantage:

```
A_i(s, a) = Q(s, a) − Σ_{u} π_i(u | τ_i) · Q(s, (a_{-i}, u))
```

where:
- `s` is the global state (in our env: the per-agent flat obs with the identity one-hot tail stripped — already implemented as `_global_obs_from_per_agent` at `bucket_brigade/training/joint_trainer.py:323-351`),
- `a = (a_1, …, a_N)` is the joint action,
- `a_{-i}` is the joint action with agent i's slot removed,
- `π_i(· | τ_i)` is agent i's policy conditioned on its local trajectory/observation,
- `Q(s, a)` is the **centralized action-value** — outputs a vector indexed by agent i's possible discrete actions, conditioned on (s, a_{-i}).

The baseline `Σ_u π_i(u | τ_i) · Q(s, (a_{-i}, u))` is the policy-weighted average over agent i's actions holding others fixed — i.e., what would happen "on average" if agent i hadn't committed to its specific action. The advantage isolates agent i's *marginal* contribution.

### Multi-dimensional action wrinkle (REQUIRES DESIGN DECISION)

This repo's action space is **multi-discrete `[10, 2, 2]` = [house, mode, signal]** per agent (see `bucket_brigade/training/joint_trainer.py:32` and `_act_all` at line 353). Standard COMA assumes a single discrete action per agent. Builder must pick one of:

1. **Factorise the counterfactual per sub-action dim.** Compute A_i,d for each of the 3 action dims independently, treating other dims as fixed. Simplest; gradient is the sum.
2. **Joint-enumerate per-agent.** Sum over the 40 joint per-agent actions (`10 × 2 × 2`) when computing the baseline. Tractable (40 forward-passes through Q is cheap) and most faithful to the COMA paper.
3. **Stochastic baseline estimate.** Sample `k` per-agent actions from π_i to approximate the sum. Highest variance, lowest compute.

**Recommended**: Option 2 (joint-enumerate). 40 entries × 4 agents × T steps is well within budget on `COMPUTE_HOST_PRIMARY` for the 50-iter × 3-seed smoke. Document the choice in the PR body so the verdict is interpretable.

### Centralized Q-critic architecture

The existing `CentralizedCritic` (`bucket_brigade/training/networks.py:210-266`) is a V-function (input: global obs → scalar). COMA needs a **Q-function**: input is `(global_obs, joint_action_one_hot_excluding_agent_i, agent_id_one_hot)` → vector of length `|A_i|` (per-agent action-value head).

Suggested structure (MLP, mirrors `CentralizedCritic` capacity):

```
Input: concat(global_obs,                  # length global_obs_dim
               flat(joint_action_one_hot),  # length N * sum(action_dims) = 4*14 = 56
               agent_id_one_hot)            # length N = 4
→ Linear(hidden=64) → ReLU → Linear(hidden=64) → ReLU
→ Linear(action_dim_d)                       # one head per sub-action dim, OR
→ Linear(40)                                 # joint-per-agent head if option 2 chosen
```

For option 2 with joint-per-agent enumeration, the cleanest pattern is a single head of size `prod(action_dims) = 40`. Builder may instead factorize as 3 heads of sizes `[10, 2, 2]` if they pick option 1.

### Wiring into the existing trainer

The current `JointPPOTrainer.collect_rollout` and update loop (`bucket_brigade/training/joint_trainer.py:393-432, ~570-700`) are PPO-style policy-gradient with GAE. COMA replaces:
- **Advantage computation**: instead of GAE on V-values, compute `A_i^COMA` from the Q-critic at each `(s_t, a_t)`.
- **Critic loss**: instead of MSE between V and returns, regress Q toward 1-step or n-step TD targets `r_t + γ · Q_target(s_{t+1}, a_{t+1})` (on-policy COMA uses the next sampled action; the paper uses TD(λ) — implementer should follow the paper).
- **Critic-target network**: COMA uses a slow-updating target Q (Polyak τ or hard copy every K steps). Add this for stability.

Keep the actor update as PPO-clip (the Foerster paper uses vanilla policy-gradient, but PPO-clip is a standard adaptation; cite this choice in PR body). The actor's policy loss becomes `−E[ log π_i(a_i|τ_i) · A_i^COMA ]` with PPO clip applied to the ratio just as today.

### Hyperparameters (starting point)

Match the MAPPO sweep on PR #232 so the comparison is apples-to-apples:
- `lr=3e-4`, `gamma=0.99`, `clip_epsilon=0.2`, `value_coef=0.5`, `entropy_coef=0.01`
- Q-critic lr: same `3e-4` initially; tune if unstable
- Target network: hard copy every 200 update steps (Foerster default)
- Rollout: 2048 steps × 50 iters × 3 seeds, matching #232

### Success criteria

- [ ] **Implementation correctness**: A unit test that recovers the COMA baseline on a fixed `(s, π_i, Q)` against a hand-computed reference (see `tests/test_joint_trainer.py` style)
- [ ] **Convergence on smoke**: COMA runs 50 iters on `minimal_specialization` and `default` without NaN / divergence, on 3 seeds each
- [ ] **Verdict logged per pre-reg ladder**: `gap_closed` computed against the same baseline as PR #232 (random baseline 308 reward; #174 target 320). Verdict tier (≥0.5 / 0.25–0.5 / <0.25) recorded in the results notebook.
- [ ] **Honest comparison**: a paired IPPO vs COMA vs MAPPO run on the same seeds (extend `analyze_231.py`), since the active question is "does counterfactual credit > shared baseline?"
- [ ] Final result documented in `research_notebook/2026-MM-DD_issue284_coma_verdict.md`

### Dependencies

- [x] #270 (BC-init basin-trap probe) — closed; verdict was anti-attractor / basin-trap mixed (#272 verdict notebook). Confirms COMA still on the table per the user's downgrade comment.
- [ ] **Human re-promotion** — see Priority note above. The author has explicitly de-prioritised COMA in favor of #286 / #288. This issue should NOT auto-graduate to `loom:issue` even after curation.

### Risks and mitigations

1. **Q-critic variance**: counterfactual baselines can have higher variance than V-baselines, especially with multi-discrete actions. Mitigation: target network, normalised advantages (already on by default in `JointPPOTrainer.update`), and the joint-enumeration baseline (option 2) which is exact rather than sampled.
2. **Action-space combinatorics**: `prod([10,2,2]) = 40` is fine; if we later add bigger discrete dims this approach won't scale and we'd need a factored Q (option 1).
3. **Comparison fairness**: MAPPO had a V-critic; COMA has a Q-critic with ~10× more output dim. To match capacity, give MAPPO a comparable parameter budget when re-running the paired sweep, or note the asymmetry in the PR body.

## Affected Files

- `bucket_brigade/training/networks.py` — add `CentralizedQCritic` class alongside the existing `CentralizedCritic` (line 210). Mirror trunk/capacity; differ in output shape and input concatenation.
- `bucket_brigade/training/joint_trainer.py` — add `coma=True` flag analogous to `centralized_critic`. Branch in `__init__` (around line 213), `_act_all` (line 353), `collect_rollout` (line 393), and the update loop (~570–700) to: (a) instantiate Q-critic + target, (b) compute counterfactual advantages from Q at update time, (c) regress Q toward TD targets. Reuse `_global_obs_from_per_agent` (line 323) for the global state input.
- `experiments/p3_specialization/train.py` — add `--coma` CLI flag mirroring `--centralized-critic` (line 803) and `cfg.coma` (line 81). Update the trainer call (line 654, 877) to pass it. Forbid both `--coma` and `--centralized-critic` simultaneously.
- `tests/test_joint_trainer.py` — add a unit test verifying the COMA counterfactual baseline is computed correctly on a fixed input (hand-computed reference).
- `experiments/p3_specialization/analyze_231.py` — generalise to a 3-arm comparison (IPPO / MAPPO / COMA), or fork to `analyze_284.py`. Per-scenario gap-closure + verdict tier per the pre-reg ladder.
- `experiments/p3_specialization/diagnostics/results/issue284_coma/` — new directory for the smoke-sweep results.
- `research_notebook/2026-MM-DD_issue284_coma_verdict.md` — protocol, references, full results, implications. Mirrors the style of `2026-05-16_issue231_mappo_phase2.md` (referenced in PR #232).

## Test Plan

- [ ] **Unit test**: hand-compute COMA advantage on a 2-agent, 2-action toy example with known Q and π; verify `A_i^COMA` matches to 1e-6.
- [ ] **Smoke**: `uv run python experiments/p3_specialization/train.py --scenario minimal_specialization --lambda-red 0.0 --seed 0 --num-iterations 5 --rollout-steps 512 --coma --output-dir /tmp/coma_smoke` completes without NaN / shape error.
- [ ] **Full sweep**: 50-iter × 2048-step × 3-seed × `{minimal_specialization, default}` × `{ippo, mappo, coma}` on `COMPUTE_HOST_PRIMARY` (Mac Studio) in tmux. Wall-clock budget: ≤30 min based on #232 timing.
- [ ] **Verdict**: `analyze_284.py` produces a 2-scenario × 3-arm table with gap_closed, action entropy, and the verdict tier per the pre-reg ladder.
- [ ] **Regression guard**: the existing `centralized_critic=True` and IPPO paths continue to pass `pytest tests/test_joint_trainer.py` unchanged (no behavioural drift when `coma=False`).
- [ ] **Edge cases**: (a) `coma=True` with `redundancy_coef>0` raises `ValueError` (matches the MAPPO incompatibility pattern at line 236); (b) `coma=True` with `centralized_critic=True` raises `ValueError`; (c) joint-action enumeration handles single-agent edge gracefully (N=1 → COMA collapses to advantage = Q − V).

## Related issues

- **Same family (algorithm-side credit assignment / CTDE)**: #208 (MAPPO impl, closed), #231 (MAPPO Phase-2 sweep, closed), #287 (LOLA — opponent-awareness), #289 (HCA — retrospective credit). These three plus this one are the Foerster-et-al cluster.
- **Sibling Phase-3 interventions**: #283 (potential-based shaping), #265 (dense progress signal).
- **Decision framework**: #193 (reward-shaping decision frame — lists CTDE/COMA/QMIX as algorithm-side option 6).
- **Recent priority pivot**: #270 (BC-init, closed → #272 verdict), #271 (best-of-N, closed → 2026-05-17 verdict notebook). Author's comment on this issue documents the downgrade.

**Comment by rjwalters** [comment]
## Re-prioritization post-#271

#271 falsified the active-anti-attractor reading. COMA's counterfactual baseline addresses credit-assignment dilution in the gradient — which is a 'direction' fix. Since the new thesis is 'gradient locality, not direction,' COMA's leverage decreases.

Still worth running if cheaper basin-escape experiments (#288 PBT, #286 macro-actions) don't help, but no longer Phase-2-priority. See `research_notebook/2026-05-17_issue271_best_of_n_verdict.md` for the pivot.

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete. Enhancement summary:

- Added COMA counterfactual-advantage math (Foerster 2018) with the per-agent baseline formula and explicit treatment of our multi-discrete `[10, 2, 2]` action space (3 design options; recommended: joint-enumerate, 40 entries).
- Specified the new `CentralizedQCritic` architecture and how it differs from the existing `CentralizedCritic` V-function (`bucket_brigade/training/networks.py:210`).
- Wired the integration points into `JointPPOTrainer` (`bucket_brigade/training/joint_trainer.py:213, 323, 393, ~570-700`) and the experiment runner (`experiments/p3_specialization/train.py:81, 803, 877`).
- Added Affected Files, Test Plan, and Success Criteria sections (Builder quality-check required).
- Cross-referenced the Phase-3 sibling cluster: #287 (LOLA), #289 (HCA), #283 (potential shaping), and the MAPPO predecessor #208 / #231 / PR #232.
- Documented the priority caveat: per the author's comment, COMA is **de-prioritised** in favor of #286 / #288 after the #271 thesis pivot. This issue should not auto-graduate; it is curated documentation pending human re-promotion.

Marking `loom:curated`. Per Curator role, NOT adding `loom:issue` — that gate is the user's.
